Web/Forma.hs

```diff
@@ -362,6 +364,53 @@ field' = field @name check
     check :: a -> ExceptT () m a
     check = return
 
+
+-- | Transform a form by applying another validator on its result.
+-- 
+-- You can layer checkers around sub-forms, like so:
+--
+-- > tupleValuesMatch err (a, b) = do
+-- >   if a == b
+-- >     then return a
+-- >     else throwError err
+-- >
+-- > createNewPasswordForm = 
+-- >   addValidation
+-- >     ((,) <$> field @"password" notEmpty <*> field @"password_confirmation" notEmpty)
+-- >     (tupleValuesMatch (fieldError @"password_confirmation" "Passwords don't match!"))
+-- 
+-- Note that you must specify the (field) name on which to add a validation error message
+-- in case the check fails.
+--
+addValidation :: forall (names :: [Symbol]) m a b.
```

Comment:
OK, one thing needs to be clarified. `FieldError` is a rather internal type. We allow to create it for use in callback because we want to allow the user to say something about arbitrary fields. In this case however, a more handy API is achieveable:

```haskell
withCheck :: forall (name :: Symbol) (names :: [Symbol]) m e s a.
  ( KnownSymbol name
  , InSet name names
  , Monad m
  , ToJSON e )
  => (s -> ExceptT e m a)
  -> FormParser names m s
  -> FormParser names m a
withCheck check (FormParser f) = FormParser $ \v -> do
  let name = pick @name @names
  r <- f v
  case r of
    Succeeded x -> do
      res <- runExceptT (check x)
      return $ case res of
        Left verr ->
          ValidationFailed (mkFieldError name verr)
        Right y ->
          Succeeded y
    ValidationFailed e ->
      return (ValidationFailed e)
    ParsingFailed msg ->
      return (ParsingFailed msg)
```

Note that we use `e` here as error message, which is anything JSON-encodable. In the examples it's just `Text`. Thus `fieldError` is not necessary. I also propose we call the new helper `withCheck` because "check" is the term that is used in the library. It's also shorter and the "with" part reflects that we use some computation altering its handling a bit.

The example then becomes:

```haskell
type SignupFields = '["username", "password", "password_confirmation"]

data SignupForm = SignupForm
  { signupUsername :: Text
  , signupPassword :: Text
  }

tupleValuesMatch :: Monad m => (Text, Text) -> ExceptT Text m Text
tupleValuesMatch (a,b) =
  if a == b
    then return a
    else throwError "Passwords don't match!"

signupForm :: Monad m => FormParser SignupFields m SignupForm
signupForm = SignupForm
  <$> field @"username" notEmpty
  <*> withCheck @"password_confirmation" tupleValuesMatch
        ((,) <$> field @"password" notEmpty
             <*> field @"password_confirmation" notEmpty)
```

I think this order of arguments is also more handy.

----

P.S. You have some trailing whitespace and empty lines, which is something I try to avoid. It's not so important though, you can leave them if it's hard to turn on highlighting of trailing whitespace in your editor. I'll then amend your commit and apply it with some clean up instead of merging.